ratis-server-api/src/main/java/org/apache/ratis/statemachine/TransactionContext.java

```diff
@@ -96,9 +97,28 @@ public interface TransactionContext {
   /**
    * Returns the committed log entry
    * @return the committed log entry
+   *
+   * @deprecated use {@link TransactionContext#getLogEntryRef()} instead.
    */
+  @Deprecated
   LogEntryProto getLogEntry();
 
+  /**
+   * Returns the {@link ReferenceCountedObject} that references to committed log entry.
+   *
+   * Clients of this method may call {@link ReferenceCountedObject#get()} to access the log entry immediately inside
+   * the scope of {@link StateMachine#applyTransaction}.
+   *
+   * If the log entry is needed after {@link StateMachine#applyTransaction} returns, e.g. for asynchronous computation
+   * or caching, the client code must invoke {@link ReferenceCountedObject#retain()} and
+   * {@link ReferenceCountedObject#release()}.
+   *
+   * @return Reference to the committed log entry.
+   */
```

Comment:
The javadoc is good.  Some minor suggestion below:
```java
  /**
   * Get a {@link ReferenceCountedObject} to the committed log entry.
   *
   * It is safe to access the log entry by calling {@link ReferenceCountedObject#get()}
   * (without {@link ReferenceCountedObject#retain()})
   * inside the scope of {@link StateMachine#applyTransaction}.
   *
   * If the log entry is needed after {@link StateMachine#applyTransaction} returns,
   * e.g. for asynchronous computation or caching,
   * the caller must invoke {@link ReferenceCountedObject#retain()} and {@link ReferenceCountedObject#release()}.
   *
   * @return a reference to the committed log entry if it exists; otherwise, returns null.
   */
```